src/asl.ts

```diff
@@ -497,12 +528,63 @@ export class ASL {
     return this.stateNames.get(stmt)!;
   }
 
-  public execute(stmt: Stmt): States {
+  private rewriteNext(
+    state: State,
+    scope: string,
+    scopeNames: Set<string>
+  ): State {
+    const updateIfInScopeNames = (next: string) =>
+      scopeNames.has(next) ? this.scopedStateName(scope, next) : next;
+    if (state.Type === "Choice") {
+      return {
+        ...state,
+        Choices: state.Choices.map((choice) => ({
+          ...choice,
+          Next: updateIfInScopeNames(choice.Next),
+        })),
+        Default: state.Default
+          ? updateIfInScopeNames(state.Default)
+          : undefined,
+      };
+    } else if (!("Next" in state)) {
+      return state;
+    }
+    return {
+      ...state,
+      Next: state.Next ? updateIfInScopeNames(state.Next) : state.Next,
+    };
+  }
+
+  private scopedStateName(scope: string, localName: string) {
+    if (localName === "default") {
+      return scope;
+    }
+    return `${localName}__${scope}`;
+  }
+
+  private normalizeStates(name: string, states: State | States) {
+    if ("Type" in states) {
+      return {
+        [name]: states as State,
+      };
+    } else {
```

Comment:
Is there an edge case here when the name of a state is `Type`?

Reply:
No longer an issue. Sub-states now contains a property called  `states` with the dynamic keys.